### PR TITLE
Compact decoder state after finished-source removal

### PR DESCRIPTION
**Pull request by gabrieldernbach**
:robot: gabriel-agent

## Summary
- Avoid reindexing source memory and context-attention caches during ordinary beam reordering.
- Compact source-dependent state only when finished source rows are removed, while continuing to reorder self-attention state for ordinary beam changes.

## Correctness
- Keep the memory bank, lengths, padding masks, source map, and decoder caches aligned during compaction.
- Add focused tests comparing selective mapping with the legacy full-mapping path and verifying translator compaction behavior.

## Validation
- `python3 -m pytest -q tests/test_beam_state_mapping.py` — 7 passed.
- `pre-commit run --all-files` — passed.

## Benchmark
A paired RTX A6000 benchmark with six measurements per side reduced median core search time from `208.096 s` to `205.210 s` (**1.39% faster, 1.0141x**). The PR head was faster in all six pairs, with identical normalized outputs, `24/32` solved targets, and `116` model calls. Peak allocated memory was flat; median reserved memory fell by `74 MiB`.

See the [full protocol, dispersion, and environment details](https://github.com/microsoft/retrochimera/pull/38#issuecomment-5637112360). Treat the approximately 1.4% gain as workload-specific.

### CODE REVIEW

**Comment by gabrieldernbach** [comment]
:robot: gabriel-agent

## Paired GPU benchmark

Benchmarked public base `34995e6fd76b` against PR head `e0fb5a54eb6b` on September 11, 2026.

| Metric | Base | PR head | Change |
|---|---:|---:|---:|
| Core search wall time, median | 208.096 s | 205.210 s | **-2.887 s (1.39% faster, 1.0141x)** |
| Wrapper wall time, median | 216.569 s | 213.662 s | -2.908 s |
| Throughput | 0.153775 targets/s | 0.155938 targets/s | **+1.41%** |
| Peak allocated memory, median | 3.716 GiB | 3.716 GiB | +0.4 MiB |
| Peak reserved memory, median | 5.584 GiB | 5.512 GiB | -74.0 MiB |

- **Pairing:** one warmup per side, then six measurements per side in `ABBABAABBAAB` order.
- **Consistency:** the PR head was faster in **6/6** paired measurements; median paired saving was **3.059 s**. The two-sided exact sign-test value is `0.03125`; sample size remains small.
- **Dispersion:** base median/MAD/IQR/range = `208.096/3.434/6.121/9.653 s`; head = `205.210/1.990/4.665/6.561 s`.
- **Correctness:** normalized target summaries were identical across all 12 measured runs (`4f251f9c554d…`): `24/32` solved targets and `116` total reaction-model calls.
- **Workload:** 32 fixed targets (`73842c4dd88b…`), top-50, at most 200 iterations and 4 reaction-model calls per target, no model cache.
- **Environment:** NVIDIA RTX A6000, CUDA device 2, PyTorch `2.2.2+cu121`, Syntheseus `0.8.0`, RDKit `2023.09.6`. The runner reported no timeout or external GPU process.

**Conclusion:** this is a small but consistent, output-exact improvement on the audited workload. Treat the approximately 1.4% gain as workload-specific rather than a broad throughput claim.

File: retrochimera/opennmt/decode/beam_search.py [hunk 1]
```diff
@@ -216,11 +219,13 @@ def update_finished(self):
         self.is_finished = self.is_finished.to("cpu")
         self.top_beam_finished |= self.is_finished[:, 0].eq(1)
         predictions = self.alive_seq.view(_B_old, self.beam_size, step)
+        alive_attn = cast(torch.Tensor, self.alive_attn)
```
Comment (lgeiger) [inline retrochimera/opennmt/decode/beam_search.py]: Why do we need these casts? Are they necessary to make CI happy? If not, I'd prefer to keep it as is.

File: retrochimera/opennmt/decode/translator.py [hunk 2]
```diff
@@ -188,17 +188,24 @@ def _translate_batch_with_strategy(
...
-        src_pad_len = memory_bank_bt.size(1)
+        if isinstance(memory_bank, tuple):
+            memory_bank_bt = tuple(x.transpose(0, 1).contiguous() for x in memory_bank)
+            src_pad_len = memory_bank_bt[0].size(1)
+            memory_device = memory_bank_bt[0].device
+        else:
+            tensor_memory_bank_bt = memory_bank.transpose(0, 1).contiguous()
+            memory_bank_bt = tensor_memory_bank_bt
+            src_pad_len = tensor_memory_bank_bt.size(1)
+            memory_device = tensor_memory_bank_bt.device
```
Comment (lgeiger) [inline retrochimera/opennmt/decode/translator.py]: No need for an extra `tensor_memory_bank_bt` variable
```suggestion
            memory_bank_bt = memory_bank.transpose(0, 1).contiguous()
            src_pad_len = memory_bank_bt.size(1)
            memory_device = memory_bank_bt.device
```

File: retrochimera/opennmt/decode/translator.py [hunk 3]
```diff
@@ -241,14 +248,15 @@ def _translate_batch_with_strategy(
 
             decode_strategy.advance(log_probs, attn)
             any_finished = decode_strategy.is_finished.any()
+            source_rows_compacted = False
             if any_finished:
-                decode_strategy.update_finished()
+                source_rows_compacted = bool(decode_strategy.update_finished())
```
Comment (lgeiger) [inline retrochimera/opennmt/decode/translator.py]:
```suggestion
                source_rows_compacted = decode_strategy.update_finished()
```

File: retrochimera/opennmt/decode/translator.py [hunk 4]
```diff
@@ -263,9 +271,12 @@ def _translate_batch_with_strategy(
                 if src_map is not None:
                     src_map = src_map.index_select(1, select_indices)
 
-            if parallel_paths > 1 or any_finished:
+            if select_indices is not None:
```
Comment (lgeiger) [inline retrochimera/opennmt/decode/translator.py]: Should we not keep the `parallel_paths=1` fast path, or am I missing something?
```suggestion
            if select_indices is not None and (parallel_paths > 1 or source_rows_compacted):
```

**Review by lgeiger — CHANGES_REQUESTED** [review]
Great! I only have a few small comments other then that this looks good.

Please, also add a entry to the [CHANGELOG.md](https://github.com/microsoft/retrochimera/blob/main/CHANGELOG.md).

File: tests/test_beam_state_mapping.py [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+
+    def update_finished(self) -> bool:
+        if self.step == 0:
+            return False
+        self.active_rows = 2
+        self.batch_offset = torch.tensor([1])
+        return True
+
+
+def test_translator_maps_source_rows_only_after_compaction() -> None:
```
Comment (lgeiger) [inline tests/test_beam_state_mapping.py]: Not sure if this test is actually useful. We overwrite the decoder and the strategy, so not sure whether this would actually catch any real issues, or am I missing something?

File: retrochimera/opennmt/decode/beam_search.py (hunk 1 above)
Comment (gabrieldernbach) [reply retrochimera/opennmt/decode/beam_search.py]: The cast was introduced to satisfy mypy. The CI would fail otherwise. 

I did not check for redundant casts, and chose _not_ to update the the mypy contract in parallel to optimization. I would rather like to address this in a broader typing clean up. 

If we want to avoid that cast, we can alternatively update the declaration of the decoder state types. I currently default to the current cast "cast"

Comment (lgeiger) [reply retrochimera/opennmt/decode/beam_search.py]: Happy, to keep any casts if necessary to make mypy happy. But this one doesn't look correct to me.
In the next line we again explicitly check for `if self.alive_attn is not None`, so I would expect mypy to either already correctly infer it or if not adding this type should fix it:
https://github.com/microsoft/retrochimera/blob/bf5ec59eec9ef32911d5167bf7528e3648622be1/retrochimera/opennmt/decode/decoder_strategy.py#L183
```python
self.alive_attn: Optional[torch.Tensor] = None
```

Or am I missing something?

Comment (gabrieldernbach) [reply retrochimera/opennmt/decode/beam_search.py]: Adopted it, thanks

File: retrochimera/opennmt/decode/translator.py (hunk 2 above)
Comment (gabrieldernbach) [reply retrochimera/opennmt/decode/translator.py]: Similar to the above, this was to satisfy mypy. Adjusted the constraint instead and dropped the extra memroy_bank

**Comment by gabrieldernbach** [comment]
:robot: gabriel-agent

Addressed all five inline review points:

- Declared optional tensor state explicitly and removed the casts, retaining the existing attention guard and adding an initialization assertion for memory lengths.
- Removed `tensor_memory_bank_bt`; an explicit tensor-or-tuple annotation now supports the simpler assignment.
- Removed the redundant `bool(...)` around `update_finished()`.
- Restored the `parallel_paths=1` fast path, with regression coverage.
- Replaced the scripted-strategy test with real `BeamSearch`/`TransformerDecoder` integration coverage. It compares predictions, scores, and attention against full state mapping and verifies cache reuse and correct compaction.

All 13 focused tests and local pre-commit checks pass. The changelog entry and CLA acknowledgment are still outstanding.

File: retrochimera/opennmt/decode/translator.py (hunk 3 above)
Comment (gabrieldernbach) [reply retrochimera/opennmt/decode/translator.py]: accepted, this was unnecessary

File: retrochimera/opennmt/decode/translator.py (hunk 4 above)
Comment (gabrieldernbach) [reply retrochimera/opennmt/decode/translator.py]: applied

File: tests/test_beam_state_mapping.py (hunk 5 above)
Comment (gabrieldernbach) [reply tests/test_beam_state_mapping.py]: replaced test

File: tests/test_beam_state_mapping.py [hunk 6]
```diff
@@ -0,0 +1,440 @@
...
+    assert decoder.calls == expected_calls
+    assert observed_rows == [2 * parallel_paths, 2 * parallel_paths, parallel_paths]
+    assert results["predictions"][0][0].tolist() == [4, 2]
+    assert results["predictions"][1][0].tolist() == [4, 4, 2]
+    assert beam.done
+
+
+@pytest.mark.parametrize("tuple_memory", [False, True])
+@torch.inference_mode()
+def test_translator_real_decoder_matches_full_mapping(monkeypatch, tuple_memory: bool) -> None:
```
Comment (lgeiger) [inline tests/test_beam_state_mapping.py:327]: I'm still not quite sure whether I understand what this test is doing. Could you elaborate?

**Review by lgeiger — COMMENTED** [review]
Please, also add a entry to the [CHANGELOG.md](https://github.com/microsoft/retrochimera/blob/main/CHANGELOG.md) and agree with the CLA bot, other than that this looks good.

**Review by lgeiger — APPROVED** [review]

**Comment by gabrieldernbach** [comment]
@microsoft-github-policy-service agree company="Microsoft"

File: tests/test_beam_state_mapping.py (hunk 6 above)
Comment (gabrieldernbach) [reply tests/test_beam_state_mapping.py]: I wanted two things checked: that skipping source-cache copies preserves decoding results, and that we only remove source rows when the entire source finishes—not when an individual beam finishes.

The decoder tests already checked that  map_state  could selectively reorder caches. What they didn’t cover was the translator deciding when to do that and supplying the right row selection. This is the gap the tests should fill.

It compares against full state mapping for correctness and checks cache identity to ensure we actually avoid the redundant copies. The controlled encoder inputs and EOS timing make that particular transition reproducible without needing a trained checkpoint.

I’ve split the checks and added a few comments. 

I can iterate on the tests, or if you prefer drop them as well. They originate from a somewhat cautious mutation testing request.
